In [22]:
import matplotlib.pyplot as plt
import pickle
import numpy as np
import csv
import torch
import pandas as pd
import scipy.sparse as sp

In [23]:
import sys
import os

module_path = os.path.abspath(os.path.join('..'))
if module_path not in sys.path:
    sys.path.append(module_path)

In [24]:
drug_links = pd.read_csv('../molecular/drug links.csv')
#drug_links = drug_links.astype({'PubChem Compound ID': int})

In [25]:
drug_links[drug_links["DrugBank ID"] == "DB00115"]

DrugBank ID            Name CAS Number          Drug Type  \
102     DB00115  Cyanocobalamin    68-19-9  SmallMoleculeDrug   

    KEGG Compound ID KEGG Drug ID  PubChem Compound ID  PubChem Substance ID  \
102           C02823       D00166           70678590.0            46509031.0   

     ChEBI ID PharmGKB ID  ... GenBank ID           DPD ID  \
102   17439.0    PA451892  ...        NaN  781; 4821; 4991   

                                      RxList Link  \
102  http://www.rxlist.com/cgi/generic/b12inj.htm   

                                        Pdrhealth Link    Wikipedia ID  \
102  http://www.pdrhealth.com/drug_info/nmdrugprofi...  Cyanocobalamin   

                                       Drugs.com Link NDC ID ChemSpider ID  \
102  http://www.drugs.com/cdi/cyanocobalamin-b12.html    NaN    21864832.0   

    BindingDB ID  TTD ID  
102          NaN     NaN  

[1 rows x 23 columns]

In [26]:
with open('../index_map/mapping_dicts/new_drug_map.pkl', 'rb') as f: 
    new_drug_map = pickle.load(f)
drugs = list(new_drug_map.keys())

In [27]:
len(drugs)

645

In [28]:
drug_links_no_nan = drug_links[~drug_links['PubChem Compound ID'].isnull()]
# drug_links_no_nan = drug_links_no_nan[drug_links_no_nan['PubChem Compound ID'].isin(drugs)]
drug_links_no_nan

DrugBank ID                                Name   CAS Number  \
5        DB00006                         Bivalirudin  128270-60-0   
13       DB00014                           Goserelin   65807-02-5   
25       DB00027                        Gramicidin D    1405-97-6   
47       DB00050                          Cetrorelix  120287-85-6   
73       DB00080                          Daptomycin  103060-53-3   
...          ...                                 ...          ...   
3802     DB14598  Edetate calcium disodium anhydrous      62-33-9   
3803     DB14600          Edetate disodium anhydrous     139-33-3   
3826     DB14632               Prednisolone tebutate    7681-14-3   
3832     DB14649               Dexamethasone acetate    1177-87-3   
3833     DB14650                Menadiol diphosphate      84-98-0   

              Drug Type KEGG Compound ID KEGG Drug ID  PubChem Compound ID  \
5     SmallMoleculeDrug              NaN       D03136           16129704.0   
13    SmallMoleculeDrug              NaN       D00573            5311128.0   
25    SmallMoleculeDrug              NaN       D04369           45267103.0   
47    SmallMoleculeDrug              NaN       D01685           25074887.0   
73    SmallMoleculeDrug           C12013       D01080           16134395.0   
...                 ...              ...          ...                  ...   
3802  SmallMoleculeDrug              NaN          NaN               6109.0   
3803  SmallMoleculeDrug              NaN       D03945               8759.0   
3826  SmallMoleculeDrug           C08182       D00982              93055.0   
3832  SmallMoleculeDrug           C08174       D07796             236702.0   
3833  SmallMoleculeDrug              NaN          NaN               8556.0   

      PubChem Substance ID  ChEBI ID  PharmGKB ID  ... GenBank ID DPD ID  \
5               46507415.0   59173.0      PA10032  ...        NaN  12945   
13              46507336.0    5523.0  PA164747674  ...        NaN  11167   
25              46507412.0       NaN     PA449808  ...        NaN   8523   
47              46505494.0   59224.0  PA164764506  ...        NaN  13224   
73              46504551.0  600103.0  PA164768820  ...        NaN  20161   
...                    ...       ...          ...  ...        ...    ...   
3802                   NaN    4757.0          NaN  ...        NaN    NaN   
3803                   NaN   64734.0          NaN  ...        NaN    NaN   
3826                   NaN    8381.0          NaN  ...        NaN    NaN   
3832                   NaN    4463.0          NaN  ...        NaN    NaN   
3833                   NaN  134713.0          NaN  ...        NaN    NaN   

                                            RxList Link Pdrhealth Link  \
5        http://www.rxlist.com/cgi/generic/angiomax.htm            NaN   
13        http://www.rxlist.com/cgi/generic/goserel.htm            NaN   
25                                                  NaN            NaN   
47    http://www.rxlist.com/cgi/generic2/cetrorelix.htm            NaN   
73       http://www.rxlist.com/cgi/generic3/cubicin.htm            NaN   
...                                                 ...            ...   
3802                                                NaN            NaN   
3803                                                NaN            NaN   
3826                                                NaN            NaN   
3832                                                NaN            NaN   
3833                                                NaN            NaN   

                         Wikipedia ID  \
5                         Bivalirudin   
13                          Goserelin   
25                       Gramicidin_D   
47                         Cetrorelix   
73                         Daptomycin   
...                               ...   
3802           Sodium_calcium_edetate   
3803  Ethylenediaminetetraacetic_acid   
3826            Prednisolone_tebutate   
3832            Dexamethaso

In [29]:
drug_links_no_nan = drug_links_no_nan[drug_links_no_nan['PubChem Compound ID'].isin(drugs)]
drug_links_no_nan

DrugBank ID             Name  CAS Number          Drug Type  \
155      DB00176      Fluvoxamine  54739-18-3  SmallMoleculeDrug   
160      DB00181         Baclofen   1134-47-0  SmallMoleculeDrug   
161      DB00182      Amphetamine    300-62-9  SmallMoleculeDrug   
165      DB00186        Lorazepam    846-49-1  SmallMoleculeDrug   
170      DB00191      Phentermine    122-09-8  SmallMoleculeDrug   
...          ...              ...         ...                ...   
1924     DB09110       Coenzyme M   3375-50-6  SmallMoleculeDrug   
1947     DB09134         Ioversol  87771-40-2  SmallMoleculeDrug   
1964     DB09153  Sodium chloride   7647-14-5  SmallMoleculeDrug   
1967     DB09156        Iopromide  73334-07-3  SmallMoleculeDrug   
2085     DB09389       Norgestrel   6533-00-2  SmallMoleculeDrug   

     KEGG Compound ID KEGG Drug ID  PubChem Compound ID  PubChem Substance ID  \
155            C07571          NaN               3404.0            46507588.0   
160               NaN       D00241               2284.0            46508181.0   
161            C07514       D07445               3007.0            46506414.0   
165               NaN       D00365               3958.0            46508468.0   
170            C07438       D05458               4771.0            46508515.0   
...               ...          ...                  ...                   ...   
1924           C03576          NaN                598.0           310265029.0   
1947              NaN       D01555               3741.0           310265049.0   
1964           C13563       D02056               5234.0           310265066.0   
1967              NaN       D01893               3736.0           310265069.0   
2085              NaN       D00954               4542.0           347827838.0   

      ChEBI ID  PharmGKB ID  ... GenBank ID             DPD ID  \
155     5138.0     PA449690  ...        NaN               1212   
160     2972.0     PA448533  ...        NaN               2104   
161   132233.0     PA448408  ...        NaN  8300; 8309; 22635   
165     6539.0     PA450267  ...        NaN               2300   
170     8080.0  PA164748099  ...        NaN        20320; 8289   
...        ...          ...  ...        ...                ...   
1924   17905.0          NaN  ...        NaN               1281   
1947   31717.0          NaN  ...        NaN         1187; 7218   
1964   26710.0     PA451382  ...        NaN    55; 9735; 11393   
1967   63578.0          NaN  ...        NaN                521   
2085    7630.0          NaN  ...        NaN         7494; 7503   

                                           RxList Link  \
155       http://www.rxlist.com/cgi/generic/fluvox.htm   
160     http://www.rxlist.com/cgi/generic/baclofen.htm   
161   http://www.rxlist.com/cgi/generic/adderallxr.htm   
165        http://www.rxlist.com/cgi/generic/loraz.htm   
170     http://www.rxlist.com/cgi/generic/phenterm.htm   
...                                                ...   
1924             http://www.rxlist.com/mesnex-drug.htm   
1947  http://www.rxlist.com/optiray-injection-drug.htm   
1964                                               NaN   
1967          http://www.rxlist.com/ultravist-drug.htm   
2085       http://www.rxlist.com/low-ogestrel-drug.htm   

                                         Pdrhealth Link     Wikipedia ID  \
155   http://www.pdrhealth.com/drug_info/rxdrugprofi...      Fluvoxamine   
160                                                 NaN         Baclofen   
161   http://www.pdrhealth.com/drug_info/rxdrugprofi...      Amphetamine   
165                                                 NaN        Lorazepam   
170                                                 NaN      Phentermine   
...                                                 ...              ...   
1924  http://www.pdr.net/pdr-consumer-monograph/mesn...       Coenzyme_M   
1947                                                NaN         Ioversol   
1964                                     

In [30]:
missing_drugs = []
drugs2 = list(drug_links_no_nan['PubChem Compound ID'])

for drug in drugs:
    if drug not in drugs2:
        missing_drugs.append(drug)


In [31]:
drug_names = pd.read_csv('../molecular/drug_names.csv')

In [32]:
drug_names

CID000000085     carnitine
0    CID000000119          GABA
1    CID000000143    leucovorin
2    CID000000158          PGE2
3    CID000000159  prostacyclin
4    CID000000191     adenosine
..            ...           ...
639  CID006435110    ivermectin
640  CID006436173     rifaximin
641  CID006447131  pimecrolimus
642  CID006918453     auranofin
643  CID009571074    cefditoren

[644 rows x 2 columns]

In [33]:
from utils.utils import get_drug_index_from_text

def sliceindex(x):
    i = 0
    for c in x:
        if c.isalpha():
            i = i + 1
            return i
        i = i + 1
        
def upperfirst(x):
    i = sliceindex(x)
    return x[:i].upper() + x[i:]

drug_names["PubChem Compound ID"] = drug_names["CID000000085"].apply(lambda x: get_drug_index_from_text(x))
drug_names["name"] = drug_names["carnitine"].apply(lambda x: upperfirst(x))
drug_names.drop(columns=["CID000000085","carnitine"], inplace=True)

In [34]:
drug_names

PubChem Compound ID          name
0                    119          GABA
1                    143    Leucovorin
2                    158          PGE2
3                    159  Prostacyclin
4                    191     Adenosine
..                   ...           ...
639              6435110    Ivermectin
640              6436173     Rifaximin
641              6447131  Pimecrolimus
642              6918453     Auranofin
643              9571074    Cefditoren

[644 rows x 2 columns]

In [35]:
common_drugs = drug_names[drug_names['PubChem Compound ID'].isin(drugs)]
common_drugs

PubChem Compound ID          name
0                    119          GABA
1                    143    Leucovorin
2                    158          PGE2
3                    159  Prostacyclin
4                    191     Adenosine
..                   ...           ...
639              6435110    Ivermectin
640              6436173     Rifaximin
641              6447131  Pimecrolimus
642              6918453     Auranofin
643              9571074    Cefditoren

[644 rows x 2 columns]

In [36]:
found_drugs = pd.merge(drug_links_no_nan, common_drugs, how='inner')

In [37]:
pd.set_option('display.max_rows', 20)
found_drugs[['DrugBank ID', 'Name', 'name', 'PubChem Compound ID']].head(400)

DrugBank ID             Name            name  PubChem Compound ID
0       DB00176      Fluvoxamine     Fluvoxamine               3404.0
1       DB00181         Baclofen        Baclofen               2284.0
2       DB00182      Amphetamine     Amphetamine               3007.0
3       DB00186        Lorazepam       Lorazepam               3958.0
4       DB00191      Phentermine     Phentermine               4771.0
..          ...              ...             ...                  ...
340     DB09110       Coenzyme M           Mesna                598.0
341     DB09134         Ioversol        Ioversol               3741.0
342     DB09153  Sodium chloride            NaCl               5234.0
343     DB09156        Iopromide       Iopromide               3736.0
344     DB09389       Norgestrel  Levonorgestrel               4542.0

[345 rows x 4 columns]

In [38]:
non_common_drugs = drug_names[drug_names['PubChem Compound ID'].isin(missing_drugs)]
non_common_drugs 

PubChem Compound ID          name
1                    143    Leucovorin
2                    158          PGE2
3                    159  Prostacyclin
4                    191     Adenosine
5                    206     Galactose
..                   ...           ...
638              6398970      Cefdinir
639              6435110    Ivermectin
641              6447131  Pimecrolimus
642              6918453     Auranofin
643              9571074    Cefditoren

[299 rows x 2 columns]

In [39]:
non_found_drugs = pd.merge(drug_links, non_common_drugs, left_on='Name', right_on='name')

In [40]:
pd.set_option('display.max_rows', 20)
non_found_drugs[['DrugBank ID', 'Name', 'PubChem Compound ID_x', 'PubChem Compound ID_y']]

DrugBank ID                     Name  PubChem Compound ID_x  \
0       DB00014                Goserelin              5311128.0   
1       DB00063             Eptifibatide                    NaN   
2       DB00104               Octreotide               448601.0   
3       DB00136               Calcitriol              5280453.0   
4       DB00162                Vitamin A               445354.0   
..          ...                      ...                    ...   
241     DB11094                Vitamin D                    NaN   
242     DB11735                Galactose              3037556.0   
243     DB11750               Clobetasol              5311051.0   
244     DB13867              Fluticasone                    NaN   
245     DB14540  Hydrocortisone butyrate                    NaN   

     PubChem Compound ID_y  
0                    47725  
1                  5311082  
2                    54374  
3                     2524  
4                     1071  
..                     ...  
241                   3249  
242                    206  
243                  32797  
244                  62924  
245                   3642  

[246 rows x 4 columns]

In [41]:
added_drugs = list(non_found_drugs['PubChem Compound ID_y'])
still_missing_drugs = list(set(missing_drugs).difference(added_drugs))
len(still_missing_drugs)

54

In [42]:
import difflib
all_drug_names = drug_links['Name'].tolist()
still_missing_drugs_df = drug_names[drug_names['PubChem Compound ID'].isin(still_missing_drugs)]
still_missing_drugs_df_list = list(still_missing_drugs_df['PubChem Compound ID'])
still_missing_drugs_check = list(set(still_missing_drugs).difference(still_missing_drugs_df_list))
still_missing_drugs_check

[85]

In [43]:
still_missing_drugs_names = list(still_missing_drugs_df['name'])
mapping_effort = {}

for drug in still_missing_drugs_names:
    mapping_effort[drug] = difflib.get_close_matches(drug, all_drug_names)
mapping_effort

{'PGE2': [],
 'Prostacyclin': ['Procyclidine', 'Metacycline', 'Tetracycline'],
 'Prostaglandin E1': [],
 'N-acetylcysteine': ['Acetylcysteine',
  'N-acetyltyrosine',
  'N-Acetylglucosamine'],
 'Cytosine arabinoside': ['Cytarabine', 'Inosine pranobex'],
 'Thyroxine': ['Tyrosine', 'Pyridoxine', 'Chloroxine'],
 'Triiodothyronine': ['Liothyronine',
  'Diiodohydroxyquinoline',
  'Dextrothyroxine'],
 'Nicotinic acid': ['Nicotinamide', 'Succinic acid', 'Lactic acid'],
 'Sulfate': ['Sulfameter', 'Sucralfate', 'Sulfacytine'],
 'Thymidine': ['Glymidine', 'Thyme', 'Thiamine'],
 'Methamphetamine': ['Metamfetamine', 'Dextroamphetamine', 'Methoxyphenamine'],
 '2-Chlorodeoxyadenosine': ['Chlorzoxazone', 'Chloroxylenol', 'Chlormezanone'],
 'Actinomycin D': ['Dactinomycin', 'Spectinomycin', 'Lincomycin'],
 'Beclomethasone': ['Betamethasone', 'Alclometasone', 'Flumethasone'],
 'Penicillin G': ['Pheneticillin', 'Penicillamine', 'Meticillin'],
 'Candesartan': ['Valsartan', 'Candesartan cilexetil', 'Vandet

In [44]:
still_missing_drugs_df

PubChem Compound ID                       name
2                    158                       PGE2
3                    159               Prostacyclin
6                    214           Prostaglandin E1
13                   581           N-acetylcysteine
14                   596       Cytosine arabinoside
..                   ...                        ...
589               160051  Colesevelam hydrochloride
590               163742        Fondaparinux sodium
592               176168                     Lambda
608              3081884         Glatiramer acetate
632              5473385              6-Thioguanine

[53 rows x 2 columns]

In [45]:
drug1 = 'DB00917'
drug2 = 'DB01240'
drug3 = 'DB00770'
drug4 = 'DB06151'
drug5 = 'DB00987'
drug6 = 'DB00451'
drug7 = 'DB00279'
drug8 = 'DB00627'
drug9 = 'DB09472'
drug10 = 'DB04485' #missing from the big df
drug11 = 'DB01577'
drug12 = 'DB00242'
drug13 = 'DB00970'
drug14 = 'DB00394'
drug15 = 'DB01053'
drug16 = 'DB13919' #missing from the big df
drug17 = 'DB00411'
drug18 = 'DB01333'
drug20 = 'DB00115' 
drug21 = 'DB00091'
drug22 = 'DB00470'
drug23 = 'DB00869'
drug24 = 'DB00158'
drug25 = 'DB00986'
drug26 = 'DB00741' 
drug27 = 'DB00332'
drug28 = 'DB00007'
drug29 = 'DB00859'
drug30 = 'DB00417' 
drug31 = 'DB00552'
drug32 = 'DB01022'
drug33 = 'DB00877'
drug34 = 'DB00209'
drug35 = 'DB00577'
drug36 = 'DB00107' 
drug37 = 'DB00603'
drug38 = 'DB01260'
drug39 = 'DB00351'
drug40 = 'DB00971' 
drug41 = 'DB01164'
drug42 = 'DB01003'
drug43 = 'DB00035'
drug44 = 'DB14509'
drug45 = 'DB00294'
drug46 = 'DB00596' 
drug47 = 'DB05271'
drug48 = 'DB01367'
drug49 = 'DB00275'
drug50 = 'DB00930' 
drug51 = 'DB00569'
drug52 = 'DB06792'
drug53 = 'DB05259'
drug54 = 'DB00352'
drug19 = 'DB00583' # for some reason not in the names list

In [46]:
still_missing_drugbanks = ['DB00917','DB01240','DB00770','DB06151','DB00987',
       'DB00451','DB00279','DB00627','DB09472','DB01577',
       'DB00242','DB00970','DB00394','DB01053','DB00411',
       'DB01333','DB00115','DB00091','DB00470',
       'DB00869','DB00158','DB00986','DB00741','DB00332',
       'DB00007','DB00859','DB00417','DB00552','DB01022',
       'DB00877','DB00209','DB00577','DB00107','DB00603',
       'DB01260','DB00351','DB00971','DB01164','DB01003',
       'DB00035','DB14509','DB00294','DB00596','DB05271',
       'DB01367','DB00275','DB00930','DB00569','DB06792',
       'DB05259','DB00352']
still_missing_drugs_nms = ['PGE2','Prostacyclin','Prostaglandin E1','N-acetylcysteine','Cytosine arabinoside',
                           'Thyroxine','Triiodothyronine','Nicotinic acid','Sulfate','Methamphetamine',
                           '2-Chlorodeoxyadenosine','Actinomycin D','Beclomethasone','Penicillin G','Carbachol',
                           'Cephradine','Vitamin B12','Cyclosporine','Delta 9-tetrahydrocannabinol','Ets',
                           'Folate','Glycopyrrolate','Cortisol','Ipratropium bromide','Leuprolide acetate',
                           'D-penicillamine','Penicillin V','DCF','Vitamin K1','Rapamycin',
                           'Trospium chloride','Valacyclovir','Synthetic oxytocin','Medroxyprogesterone','Apolar',
                           'Megestrol','Selenium sulfide','Calcium chloride','Cromolyn','DDAVP',
                           'Lithium','Implanon','Halobetasol propionate','N-0437','RSA',
                           'Olmesartan medoxomil','Colesevelam hydrochloride','Fondaparinux sodium','Lambda','Glatiramer acetate',
                           '6-Thioguanine']

In [47]:
still_missing_drugs_df

PubChem Compound ID                       name
2                    158                       PGE2
3                    159               Prostacyclin
6                    214           Prostaglandin E1
13                   581           N-acetylcysteine
14                   596       Cytosine arabinoside
..                   ...                        ...
589               160051  Colesevelam hydrochloride
590               163742        Fondaparinux sodium
592               176168                     Lambda
608              3081884         Glatiramer acetate
632              5473385              6-Thioguanine

[53 rows x 2 columns]

In [48]:
still_missing_drugs_df_minus2 = still_missing_drugs_df[still_missing_drugs_df['name'].isin(still_missing_drugs_nms)]
still_missing_drugs_df_minus2

PubChem Compound ID                       name
2                    158                       PGE2
3                    159               Prostacyclin
6                    214           Prostaglandin E1
13                   581           N-acetylcysteine
14                   596       Cytosine arabinoside
..                   ...                        ...
589               160051  Colesevelam hydrochloride
590               163742        Fondaparinux sodium
592               176168                     Lambda
608              3081884         Glatiramer acetate
632              5473385              6-Thioguanine

[51 rows x 2 columns]

In [49]:
still_missing_drugs_ids_minus2 = list(still_missing_drugs_df_minus2['PubChem Compound ID'])
still_missing_drugs_ids_nms_minus2 = [[x,y] for x,y in zip(still_missing_drugs_ids_minus2,still_missing_drugs_nms)]

In [50]:
still_missing_drugs_dict = dict(zip(still_missing_drugbanks,still_missing_drugs_ids_nms_minus2))
still_missing_drugs_dict

{'DB00917': [158, 'PGE2'],
 'DB01240': [159, 'Prostacyclin'],
 'DB00770': [214, 'Prostaglandin E1'],
 'DB06151': [581, 'N-acetylcysteine'],
 'DB00987': [596, 'Cytosine arabinoside'],
 'DB00451': [853, 'Thyroxine'],
 'DB00279': [861, 'Triiodothyronine'],
 'DB00627': [937, 'Nicotinic acid'],
 'DB09472': [1117, 'Sulfate'],
 'DB01577': [1206, 'Methamphetamine'],
 'DB00242': [1546, '2-Chlorodeoxyadenosine'],
 'DB00970': [2019, 'Actinomycin D'],
 'DB00394': [2308, 'Beclomethasone'],
 'DB01053': [2349, 'Penicillin G'],
 'DB00411': [2551, 'Carbachol'],
 'DB01333': [2673, 'Cephradine'],
 'DB00115': [2891, 'Vitamin B12'],
 'DB00091': [2909, 'Cyclosporine'],
 'DB00470': [2978, 'Delta 9-tetrahydrocannabinol'],
 'DB00869': [3154, 'Ets'],
 'DB00158': [3405, 'Folate'],
 'DB00986': [3494, 'Glycopyrrolate'],
 'DB00741': [3640, 'Cortisol'],
 'DB00332': [3746, 'Ipratropium bromide'],
 'DB00007': [3911, 'Leuprolide acetate'],
 'DB00859': [4727, 'D-penicillamine'],
 'DB00417': [4730, 'Penicillin V'],
 'DB0

Add the three drugs missing

In [51]:
still_missing_drugs_dict['DB13919'] = [2541, 'Candesartan']
still_missing_drugs_dict['DB04485'] = [1134, 'Thymidine']
still_missing_drugs_dict['DB00583'] = [85, 'No stich name']

In [52]:
drug_names[drug_names["name"] == "Levocarnitine"]

Empty DataFrame
Columns: [PubChem Compound ID, name]
Index: []

<h1>Final dataset</h1>

We will join the 241 common drugs with the 34 found by common name and then add the rest 9 by hand to match all 284 drugs!

In [53]:
final_df_1 = pd.merge(drug_links, common_drugs, how='left', on='PubChem Compound ID')
final_df_1 = final_df_1[['DrugBank ID', 'Name', 'PubChem Compound ID', 'name']]
final_df_1.rename(columns={'name':'STITCH Name'}, inplace=True)
final_df_1

DrugBank ID                              Name  PubChem Compound ID  \
0        DB00001                         Lepirudin                  NaN   
1        DB00002                         Cetuximab                  NaN   
2        DB00003                      Dornase alfa                  NaN   
3        DB00004               Denileukin diftitox                  NaN   
4        DB00005                        Etanercept                  NaN   
...          ...                               ...                  ...   
3878     DB14759                        Red pepper                  NaN   
3879     DB14762                      Risankizumab                  NaN   
3880     DB14763                       Cycloguanil                  NaN   
3881     DB14764  Dengue tetravalent vaccine, live                  NaN   
3882     DB15274                 Pertussis vaccine                  NaN   

     STITCH Name  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
...          ...  
3878         NaN  
3879         NaN  
3880         NaN  
3881         NaN  
3882         NaN  

[3883 rows x 4 columns]

In [54]:
final_df_1[final_df_1["PubChem Compound ID"] == 119]

DrugBank ID                     Name  PubChem Compound ID STITCH Name
1379     DB02530  gamma-Aminobutyric acid                119.0        GABA

In [55]:
final_df_2 = pd.merge(final_df_1, non_common_drugs, how='left', left_on='Name', right_on='name')
final_df_2

DrugBank ID                              Name  PubChem Compound ID_x  \
0        DB00001                         Lepirudin                    NaN   
1        DB00002                         Cetuximab                    NaN   
2        DB00003                      Dornase alfa                    NaN   
3        DB00004               Denileukin diftitox                    NaN   
4        DB00005                        Etanercept                    NaN   
...          ...                               ...                    ...   
3878     DB14759                        Red pepper                    NaN   
3879     DB14762                      Risankizumab                    NaN   
3880     DB14763                       Cycloguanil                    NaN   
3881     DB14764  Dengue tetravalent vaccine, live                    NaN   
3882     DB15274                 Pertussis vaccine                    NaN   

     STITCH Name  PubChem Compound ID_y name  
0            NaN                    NaN  NaN  
1            NaN                    NaN  NaN  
2            NaN                    NaN  NaN  
3            NaN                    NaN  NaN  
4            NaN                    NaN  NaN  
...          ...                    ...  ...  
3878         NaN                    NaN  NaN  
3879         NaN                    NaN  NaN  
3880         NaN                    NaN  NaN  
3881         NaN                    NaN  NaN  
3882         NaN                    NaN  NaN  

[3883 rows x 6 columns]

In [56]:
final_df_2[final_df_2["PubChem Compound ID_x"] == 119]

DrugBank ID                     Name  PubChem Compound ID_x STITCH Name  \
1379     DB02530  gamma-Aminobutyric acid                  119.0        GABA   

      PubChem Compound ID_y name  
1379                    NaN  NaN

In [57]:
final_df = final_df_2[(final_df_2["STITCH Name"].notna()) | (final_df_2["PubChem Compound ID_y"].notna()) |
                        (final_df_2["DrugBank ID"].isin(still_missing_drugbanks))]

In [58]:
pd.set_option('display.max_rows', 20)
final_df.reset_index()

index DrugBank ID                     Name  PubChem Compound ID_x  \
0        6     DB00007               Leuprolide                    NaN   
1       13     DB00014                Goserelin              5311128.0   
2       33     DB00035             Desmopressin                    NaN   
3       59     DB00063             Eptifibatide                    NaN   
4       83     DB00091              Ciclosporin              5284373.0   
..     ...         ...                      ...                    ...   
637   3168     DB11735                Galactose              3037556.0   
638   3172     DB11750               Clobetasol              5311051.0   
639   3409     DB13867              Fluticasone                    NaN   
640   3758     DB14509        Lithium carbonate                    NaN   
641   3770     DB14540  Hydrocortisone butyrate                    NaN   

    STITCH Name  PubChem Compound ID_y                     name  
0           NaN                    NaN                      NaN  
1           NaN                47725.0                Goserelin  
2           NaN                    NaN                      NaN  
3           NaN              5311082.0             Eptifibatide  
4           NaN                    NaN                      NaN  
..          ...                    ...                      ...  
637         NaN                  206.0                Galactose  
638         NaN                32797.0               Clobetasol  
639         NaN                62924.0              Fluticasone  
640         NaN                    NaN                      NaN  
641         NaN                 3642.0  Hydrocortisone butyrate  

[642 rows x 7 columns]

In [59]:
final_df['PubChem Compound ID_y'].fillna(final_df['PubChem Compound ID_x'], inplace=True)
final_df['STITCH Name'].fillna(final_df['name'], inplace=True)
final_df.drop(columns=['name'], inplace=True)

/home/vasilis/miniconda3/envs/tip-gpu/lib/python3.9/site-packages/pandas/core/generic.py:6392: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)
/home/vasilis/miniconda3/envs/tip-gpu/lib/python3.9/site-packages/pandas/core/frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [60]:
final_df.rename(columns={'Name':'DrugBank Name',
                        'PubChem Compound ID_x': 'DrugBank PubChem ID',
                        'PubChem Compound ID_y': 'STITCH ID'
                        }, inplace=True)
final_df

/home/vasilis/miniconda3/envs/tip-gpu/lib/python3.9/site-packages/pandas/core/frame.py:5039: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().rename(


DrugBank ID            DrugBank Name  DrugBank PubChem ID  \
6        DB00007               Leuprolide                  NaN   
13       DB00014                Goserelin            5311128.0   
33       DB00035             Desmopressin                  NaN   
59       DB00063             Eptifibatide                  NaN   
83       DB00091              Ciclosporin            5284373.0   
...          ...                      ...                  ...   
3168     DB11735                Galactose            3037556.0   
3172     DB11750               Clobetasol            5311051.0   
3409     DB13867              Fluticasone                  NaN   
3758     DB14509        Lithium carbonate                  NaN   
3770     DB14540  Hydrocortisone butyrate                  NaN   

                  STITCH Name  STITCH ID  
6                         NaN        NaN  
13                  Goserelin    47725.0  
33                        NaN        NaN  
59               Eptifibatide  5311082.0  
83                        NaN  5284373.0  
...                       ...        ...  
3168                Galactose      206.0  
3172               Clobetasol    32797.0  
3409              Fluticasone    62924.0  
3758                      NaN        NaN  
3770  Hydrocortisone butyrate     3642.0  

[642 rows x 5 columns]

Add the missing drugs with DB13919 and DB04485, named 'Candesartan' and 'Thymidine', as well as the drug 'Levocarnitine' with DB00583 which is missing from the names csv

In [61]:
Candesartan = pd.Series({'DrugBank ID': 'DB13919', 'DrugBank Name': 'Candesartan'})
Thymidine = pd.Series({'DrugBank ID': 'DB04485', 'DrugBank Name': 'Thymidine'})
Levocarnitine = pd.Series({'DrugBank ID': 'DB00583', 'DrugBank Name': 'Levocarnitine'})

In [62]:
final_df = final_df.append(Candesartan, ignore_index=True)
final_df = final_df.append(Thymidine, ignore_index=True)
final_df = final_df.append(Levocarnitine, ignore_index=True)
final_df

DrugBank ID            DrugBank Name  DrugBank PubChem ID  \
0       DB00007               Leuprolide                  NaN   
1       DB00014                Goserelin            5311128.0   
2       DB00035             Desmopressin                  NaN   
3       DB00063             Eptifibatide                  NaN   
4       DB00091              Ciclosporin            5284373.0   
..          ...                      ...                  ...   
640     DB14509        Lithium carbonate                  NaN   
641     DB14540  Hydrocortisone butyrate                  NaN   
642     DB13919              Candesartan                  NaN   
643     DB04485                Thymidine                  NaN   
644     DB00583            Levocarnitine                  NaN   

                 STITCH Name  STITCH ID  
0                        NaN        NaN  
1                  Goserelin    47725.0  
2                        NaN        NaN  
3               Eptifibatide  5311082.0  
4                        NaN  5284373.0  
..                       ...        ...  
640                      NaN        NaN  
641  Hydrocortisone butyrate     3642.0  
642                      NaN        NaN  
643                      NaN        NaN  
644                      NaN        NaN  

[645 rows x 5 columns]

In [63]:
for drugbank, stitch in still_missing_drugs_dict.items():
    
    final_df.loc[final_df['DrugBank ID'] == drugbank, 'STITCH Name'] = stitch[1]
    final_df.loc[final_df['DrugBank ID'] == drugbank, 'STITCH ID'] = stitch[0]

In [64]:
pd.set_option('display.max_rows', 20)
final_df

DrugBank ID            DrugBank Name  DrugBank PubChem ID  \
0       DB00007               Leuprolide                  NaN   
1       DB00014                Goserelin            5311128.0   
2       DB00035             Desmopressin                  NaN   
3       DB00063             Eptifibatide                  NaN   
4       DB00091              Ciclosporin            5284373.0   
..          ...                      ...                  ...   
640     DB14509        Lithium carbonate                  NaN   
641     DB14540  Hydrocortisone butyrate                  NaN   
642     DB13919              Candesartan                  NaN   
643     DB04485                Thymidine                  NaN   
644     DB00583            Levocarnitine                  NaN   

                 STITCH Name  STITCH ID  
0         Leuprolide acetate     3911.0  
1                  Goserelin    47725.0  
2                      DDAVP    27991.0  
3               Eptifibatide  5311082.0  
4               Cyclosporine     2909.0  
..                       ...        ...  
640                  Lithium    28112.0  
641  Hydrocortisone butyrate     3642.0  
642              Candesartan     2541.0  
643                Thymidine     1134.0  
644            No stich name       85.0  

[645 rows x 5 columns]

In [65]:
# final_df.to_pickle('../molecular/drugbank/drugbank_stitch_full.pkl')